# Problem Statement:
### Predict time interval of an HVAC system will overheat using `Machine Learninng`

In [3]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.metrics import roc_auc_score
import matplotlib.pyplot as plt
from xgboost import XGBClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout, Bidirectional
from collections import Counter
from sklearn.utils import resample
from sklearn.utils import class_weight
from sklearn.metrics import precision_recall_curve
from sklearn.metrics import f1_score
import joblib

In [4]:
!pip install tensorflow


In [5]:
from google.colab import files
uploaded = files.upload()


Saving HVAC.csv to HVAC.csv


In [6]:
df = pd.read_csv("HVAC.csv")

In [7]:
df.head()

Timestamp   T_Supply   T_Return  SP_Return  T_Saturation  \
0  2019-10-15 00:00:00+02:00  19.859999  20.469999       18.5         19.02   
1  2019-10-15 00:15:00+02:00  19.855000  20.430000       18.5         19.02   
2  2019-10-15 00:30:00+02:00  19.850000  20.410000       18.5         19.02   
3  2019-10-15 00:45:00+02:00  19.840000  20.379999       18.5         19.08   
4  2019-10-15 01:00:00+02:00  19.830000  20.350000       18.5         19.08   

   T_Outdoor  RH_Supply  RH_Return  RH_Outdoor  Energy  Power  
0  20.299999  71.110001  58.919998        79.5     0.0    0.0  
1  20.299999  71.320000  59.000000        82.0     0.0    0.0  
2  20.299999  71.470001  59.109997        79.5     0.0    0.0  
3  20.299999  71.439995  59.309998        77.0     0.0    0.0  
4  20.299999  71.580002  59.559998        79.5     0.0    0.0

In [8]:
df.tail()

Timestamp   T_Supply   T_Return  SP_Return  \
33883  2021-04-14 22:45:00+02:00  19.539999  20.004999       20.5   
33884  2021-04-14 23:00:00+02:00  19.520000  19.949999       20.5   
33885  2021-04-14 23:15:00+02:00  19.430000  19.955000       20.5   
33886  2021-04-14 23:30:00+02:00  19.420000  19.920000       20.5   
33887  2021-04-14 23:45:00+02:00  19.420000  19.900000       20.5   

       T_Saturation  T_Outdoor  RH_Supply  RH_Return  RH_Outdoor  Energy  \
33883     19.619999       14.7  39.020000  27.930000        57.0     0.0   
33884     19.539999       13.7  39.020000  28.090000        57.0     0.0   
33885     19.420000       13.7  39.399998  27.930000        57.0     0.0   
33886     19.400000       13.7  39.599998  28.039999        57.0     0.0   
33887     19.400000       13.7  39.599998  28.150000        57.0     0.0   

       Power  
33883    0.0  
33884    0.0  
33885    0.0  
33886    0.0  
33887    0.0

In [9]:
df.shape

(33888, 11)

In [10]:
df.isnull().sum()

Timestamp       0
T_Supply        0
T_Return        0
SP_Return       0
T_Saturation    0
T_Outdoor       0
RH_Supply       0
RH_Return       0
RH_Outdoor      0
Energy          0
Power           0
dtype: int64

In [11]:
df.isna().sum()

Timestamp       0
T_Supply        0
T_Return        0
SP_Return       0
T_Saturation    0
T_Outdoor       0
RH_Supply       0
RH_Return       0
RH_Outdoor      0
Energy          0
Power           0
dtype: int64

##  HVAC Dataset Column Descriptions

| **Column Name** | **Meaning** |
|------------------|-------------|
| `Timestamp` | The date and time when the sensor readings were taken. This lets us organize the data as a time series (hourly, minutely, etc.). |
| `T_Supply` | Temperature of the air that the HVAC system is blowing into the room/home (cooled or heated air). In cooling mode, it should be cold — around 16–20°C. |
| `T_Return` | Temperature of the air coming back into the HVAC system from the room. It tells us how warm or cool the room currently is. |
| `SP_Return` | The set point temperature for the return air — basically, the target temperature set on the thermostat (e.g., 24°C). The HVAC system works to bring `T_Return` closer to `SP_Return`. i.e Desired room temperature (set by user/system)|
| `T_Saturation` | Temperature of the coolant/refrigerant coil inside the HVAC system. If this gets too high, it may indicate a risk of overheating or poor cooling performance. |
| `T_Outdoor` | Temperature outside the building. This is important because HVAC systems need to work harder if it's very hot (or cold) outside. |
| `RH_Supply` | the amount of moisture (water vapor) in the air, expressed as a percentage (%).RH_Supply is the humidity level of the cool/warm air coming from the HVAC system into your room.High humidity makes rooms feel warmer; too low causes dryness. Ideally balanced between 40–60%. Example : Your AC is supplying air at 18°C. That air contains some moisture (water vapor). RH_Supply = 55% → This means the air the AC is blowing has 55% humidity|
| `RH_Return` | Humidity of the air coming back from the room to the HVAC system. This reflects the moisture level inside the space. |
| `RH_Outdoor` | Humidity outside the building. Like outdoor temp, this affects how hard the HVAC system has to work. |
| `Energy` | Total energy consumed (usually in kWh) by the HVAC system over time. It increases as the system runs longer. Useful for energy efficiency analysis. |
| `Power` | Instantaneous power usage (usually in kW) — how much energy the system is using at that specific moment. Spikes may indicate strain or inefficiency. |


In [12]:
df[df['T_Supply'] > 30].shape

(58, 11)

In [13]:
def detect_mode(row):
    if row['Power'] == 0:
        return 'idle'
    elif row['T_Supply'] < row['T_Return']:
        return 'cooling'
    elif row['T_Supply'] > row['T_Return']:
        return 'heating'
    else:
        return 'unknown'

df['mode'] = df.apply(detect_mode, axis=1)


In [14]:
# Rolling mean of T_Supply for making overheat column
df['T_Supply_roll3'] = df['T_Supply'].rolling(3).mean()


In [15]:
# Overheat column generation

def check_overheat(row):
    # Only apply logic if we have valid rolling avg
    if pd.isna(row['T_Supply_roll3']):
        return 0

    # Cooling mode: supply air should be cold; if hot, it's overheating /malfunctioning
    if row['mode'] == 'cooling' and row['T_Supply_roll3'] > 27:
        return 1

    # Heating mode: supply air should be hot; if too hot, it's overheating /malfunctioning
    elif row['mode'] == 'heating' and row['T_Supply_roll3'] < 20:  #  threshold
        return 1

    # Idle or normal conditions
    else:
        return 0

df['overheat_flag'] = df.apply(check_overheat, axis=1)


In [16]:
# checking balanced or not
df[df['overheat_flag']==1].shape

(442, 14)

In [17]:
df[df['overheat_flag']!=1].shape

(33446, 14)

In [18]:
df.columns

Index(['Timestamp', 'T_Supply', 'T_Return', 'SP_Return', 'T_Saturation',
       'T_Outdoor', 'RH_Supply', 'RH_Return', 'RH_Outdoor', 'Energy', 'Power',
       'mode', 'T_Supply_roll3', 'overheat_flag'],
      dtype='object')

In [19]:
# Creating columns to findout lagging
df['T_Return_lag1'] = df['T_Return'].shift(1)
df['Power_lag1'] = df['Power'].shift(1)
df['T_Saturation_lag1'] = df['T_Saturation'].shift(1)


In [20]:
df['delta_T_Return'] = df['T_Return'] - df['T_Return'].shift(1)
df['Power_roll3'] = df['Power'].rolling(3).mean()


In [21]:
# Converting to datetime with timezone handling
df['Timestamp'] = pd.to_datetime(df['Timestamp'], utc=True)

# extracting hour and dayofweek
df['hour'] = df['Timestamp'].dt.hour
df['dayofweek'] = df['Timestamp'].dt.dayofweek


In [22]:
pd.set_option('display.max_columns', None)

In [23]:
df.head()

Timestamp   T_Supply   T_Return  SP_Return  T_Saturation  \
0 2019-10-14 22:00:00+00:00  19.859999  20.469999       18.5         19.02   
1 2019-10-14 22:15:00+00:00  19.855000  20.430000       18.5         19.02   
2 2019-10-14 22:30:00+00:00  19.850000  20.410000       18.5         19.02   
3 2019-10-14 22:45:00+00:00  19.840000  20.379999       18.5         19.08   
4 2019-10-14 23:00:00+00:00  19.830000  20.350000       18.5         19.08   

   T_Outdoor  RH_Supply  RH_Return  RH_Outdoor  Energy  Power  mode  \
0  20.299999  71.110001  58.919998        79.5     0.0    0.0  idle   
1  20.299999  71.320000  59.000000        82.0     0.0    0.0  idle   
2  20.299999  71.470001  59.109997        79.5     0.0    0.0  idle   
3  20.299999  71.439995  59.309998        77.0     0.0    0.0  idle   
4  20.299999  71.580002  59.559998        79.5     0.0    0.0  idle   

   T_Supply_roll3  overheat_flag  T_Return_lag1  Power_lag1  \
0             NaN              0            NaN         NaN   
1             NaN              0      20.469999         0.0   
2       19.855000              0      20.430000         0.0   
3       19.848333              0      20.410000         0.0   
4       19.840000              0      20.379999         0.0   

   T_Saturation_lag1  delta_T_Return  Power_roll3  hour  dayofweek  
0                NaN             NaN          NaN    22          0  
1              19.02       -0.039999          NaN    22          0  
2              19.02       -0.020000          0.0    22          0  
3              19.02       -0.030001          0.0    22          0  
4              19.08       -0.029999          0.0    23          0

In [24]:
df.columns

Index(['Timestamp', 'T_Supply', 'T_Return', 'SP_Return', 'T_Saturation',
       'T_Outdoor', 'RH_Supply', 'RH_Return', 'RH_Outdoor', 'Energy', 'Power',
       'mode', 'T_Supply_roll3', 'overheat_flag', 'T_Return_lag1',
       'Power_lag1', 'T_Saturation_lag1', 'delta_T_Return', 'Power_roll3',
       'hour', 'dayofweek'],
      dtype='object')

In [25]:
df.isna().sum()

Timestamp            0
T_Supply             0
T_Return             0
SP_Return            0
T_Saturation         0
T_Outdoor            0
RH_Supply            0
RH_Return            0
RH_Outdoor           0
Energy               0
Power                0
mode                 0
T_Supply_roll3       2
overheat_flag        0
T_Return_lag1        1
Power_lag1           1
T_Saturation_lag1    1
delta_T_Return       1
Power_roll3          2
hour                 0
dayofweek            0
dtype: int64

In [26]:
df_model = df.dropna().copy()

In [27]:
df_model.isna().sum()

Timestamp            0
T_Supply             0
T_Return             0
SP_Return            0
T_Saturation         0
T_Outdoor            0
RH_Supply            0
RH_Return            0
RH_Outdoor           0
Energy               0
Power                0
mode                 0
T_Supply_roll3       0
overheat_flag        0
T_Return_lag1        0
Power_lag1           0
T_Saturation_lag1    0
delta_T_Return       0
Power_roll3          0
hour                 0
dayofweek            0
dtype: int64

### T_Supply and T_Supply_roll3 can leak data. we will not use those feature

In [28]:
features = [
    'T_Return',
    'SP_Return',
    'T_Saturation',
    'T_Outdoor',
    'RH_Return',
    'RH_Outdoor',
    'Power',
    'Energy',
]

X = df_model[features]
y = df_model['overheat_flag']



# Train test split
split_idx = int(len(df_model) * 0.8)
X_train, X_test = X.iloc[:split_idx], X.iloc[split_idx:]
y_train, y_test = y.iloc[:split_idx], y.iloc[split_idx:]

In [29]:
X_train.head()

T_Return  SP_Return  T_Saturation  T_Outdoor  RH_Return  RH_Outdoor  \
2  20.410000       18.5     19.020000  20.299999  59.109997        79.5   
3  20.379999       18.5     19.080000  20.299999  59.309998        77.0   
4  20.350000       18.5     19.080000  20.299999  59.559998        79.5   
5  20.320000       18.5     19.100000  20.299999  59.660000        82.0   
6  20.309999       18.5     19.119999  19.299999  59.759998        85.0   

   Power  Energy  
2    0.0     0.0  
3    0.0     0.0  
4    0.0     0.0  
5    0.0     0.0  
6    0.0     0.0

In [30]:
y_train.head()

2    0
3    0
4    0
5    0
6    0
Name: overheat_flag, dtype: int64

In [31]:


model = RandomForestClassifier(class_weight='balanced', random_state=42)
model.fit(X_train, y_train)


y_pred = model.predict(X_test)
print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))



              precision    recall  f1-score   support

           0       1.00      1.00      1.00      6745
           1       0.00      0.00      0.00        33

    accuracy                           0.99      6778
   macro avg       0.50      0.50      0.50      6778
weighted avg       0.99      0.99      0.99      6778

[[6743    2]
 [  33    0]]


### Observation:
* Model is too biased toward predicting class 0.



# Lets try XGBoost

In [32]:

# Computing class imbalance ratio
scale_pos_weight = len(y_train[y_train == 0]) / len(y_train[y_train == 1])

# Defining the model
xgb = XGBClassifier(
    scale_pos_weight=scale_pos_weight,
    eval_metric='logloss',
    random_state=42,
    n_estimators=100,
    max_depth=6,
    learning_rate=0.1
)

#  Fit the model
xgb.fit(X_train, y_train)

# Make predictions
y_pred = xgb.predict(X_test)
y_probs = xgb.predict_proba(X_test)[:, 1]

# Evaluate the model

print(" Classification Report:\n", classification_report(y_test, y_pred))
print(" Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print(" AUC Score:", roc_auc_score(y_test, y_probs))


 Classification Report:
               precision    recall  f1-score   support

           0       1.00      0.98      0.99      6745
           1       0.13      0.52      0.21        33

    accuracy                           0.98      6778
   macro avg       0.56      0.75      0.60      6778
weighted avg       0.99      0.98      0.99      6778

 Confusion Matrix:
 [[6632  113]
 [  16   17]]
 AUC Score: 0.9720938068603006


* True Negatives (TN) — correctly predicted normal condition

* False Positives (FP) — wrongly predicted overheat when it wasn’t

* False Negatives (FN) — missed actual overheat events (concerning)

* True Positives (TP) — correctly predicted overheat

### Lets use LSTM

In [33]:
print("NaNs in X_train:\n", np.isnan(X_train).sum())
print("NaNs in y_train:\n", np.isnan(y_train).sum())


NaNs in X_train:
 T_Return        0
SP_Return       0
T_Saturation    0
T_Outdoor       0
RH_Return       0
RH_Outdoor      0
Power           0
Energy          0
dtype: int64
NaNs in y_train:
 0


In [34]:
def create_lstm_sequences(X, y, time_steps=5):
    Xs, ys = [], []
    for i in range(len(X) - time_steps):
        Xs.append(X[i:(i + time_steps)])
        ys.append(y[i + time_steps])
    return np.array(Xs), np.array(ys)

# Scale the data
scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X_train)

# LSTM-ready data
time_steps = 5
X_lstm, y_lstm = create_lstm_sequences(X_scaled, y_train, time_steps=time_steps)

# Splitting
split = int(0.8 * len(X_lstm))
X_train_seq, X_test_seq = X_lstm[:split], X_lstm[split:]
y_train_seq, y_test_seq = y_lstm[:split], y_lstm[split:]


In [35]:

model = Sequential([
    LSTM(64, input_shape=(X_train_seq.shape[1], X_train_seq.shape[2]), return_sequences=False),
    Dropout(0.2),
    Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

history = model.fit(X_train_seq, y_train_seq, epochs=10, batch_size=64, validation_split=0.1)


/usr/local/lib/python3.11/dist-packages/keras/src/layers/rnn/rnn.py:200: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 10s 19ms/step - accuracy: 0.9917 - loss: 0.1349 - val_accuracy: 0.9852 - val_loss: 0.0705
Epoch 2/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 4s 14ms/step - accuracy: 0.9927 - loss: 0.0378 - val_accuracy: 0.9852 - val_loss: 0.0667
Epoch 3/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 4s 9ms/step - accuracy: 0.9918 - loss: 0.0340 - val_accuracy: 0.9852 - val_loss: 0.0683
Epoch 4/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 5s 8ms/step - accuracy: 0.9925 - loss: 0.0274 - val_accuracy: 0.9839 - val_loss: 0.0730
Epoch 5/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.9927 - loss: 0.0234 - val_accuracy: 0.9839 - val_loss: 0.0775
Epoch 6/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9910 - loss: 0.0247 - val_accuracy: 0.9843 - val_loss: 0.0812
Epoch 7/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.9932 - loss: 0.0201 - val_accuracy: 0.9825 - val_loss: 0.0852
Epoch 8/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9923 - loss: 0.0221 - val_accuracy:

In [36]:
y_pred_probs = model.predict(X_test_seq).flatten()
y_pred = (y_pred_probs > 0.5).astype(int)


print("Classification Report:\n", classification_report(y_test_seq, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test_seq, y_pred))
print("AUC Score:", roc_auc_score(y_test_seq, y_pred_probs))


170/170 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step
Classification Report:
               precision    recall  f1-score   support

           0       0.96      0.98      0.97      5203
           1       0.06      0.03      0.04       218

    accuracy                           0.94      5421
   macro avg       0.51      0.50      0.50      5421
weighted avg       0.92      0.94      0.93      5421

Confusion Matrix:
 [[5104   99]
 [ 212    6]]
AUC Score: 0.8615310150989107


# Handling imbalance

In [37]:


class_weights = class_weight.compute_class_weight(
    class_weight='balanced',
    classes=np.unique(y_train_seq),
    y=y_train_seq
)
class_weights = dict(enumerate(class_weights))

history = model.fit(
    X_train_seq, y_train_seq,
    epochs=10,
    batch_size=64,
    validation_split=0.1,
    class_weight=class_weights
)


Epoch 1/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 3s 7ms/step - accuracy: 0.9004 - loss: 0.3166 - val_accuracy: 0.9299 - val_loss: 0.2021
Epoch 2/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.9485 - loss: 0.1687 - val_accuracy: 0.8631 - val_loss: 0.3103
Epoch 3/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.9442 - loss: 0.1661 - val_accuracy: 0.8893 - val_loss: 0.2925
Epoch 4/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.9483 - loss: 0.1612 - val_accuracy: 0.9308 - val_loss: 0.2209
Epoch 5/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.9481 - loss: 0.1333 - val_accuracy: 0.9295 - val_loss: 0.2262
Epoch 6/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9589 - loss: 0.1462 - val_accuracy: 0.9272 - val_loss: 0.2275
Epoch 7/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 3s 7ms/step - accuracy: 0.9586 - loss: 0.1327 - val_accuracy: 0.8875 - val_loss: 0.2910
Epoch 8/10
305/305 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9429 - loss: 0.1359 - val_accuracy: 0.

In [38]:


# Combining
X_pos = X_lstm[y_lstm == 1]
y_pos = y_lstm[y_lstm == 1]
X_neg = X_lstm[y_lstm == 0]
y_neg = y_lstm[y_lstm == 0]

X_pos_upsampled, y_pos_upsampled = resample(
    X_pos, y_pos,
    replace=True,
    n_samples=len(y_neg),
    random_state=42
)

# Concatenate
X_balanced = np.concatenate([X_neg, X_pos_upsampled])
y_balanced = np.concatenate([y_neg, y_pos_upsampled])


In [39]:


# Shuffle and split
X_train, X_test, y_train, y_test = train_test_split(
    X_balanced, y_balanced, test_size=0.2, random_state=42, shuffle=True
)


In [40]:

# Using Bidirectional LSTM
model = Sequential([
    Bidirectional(LSTM(64, return_sequences=False), input_shape=(X_train.shape[1], X_train.shape[2])),
    Dropout(0.3),
    Dense(32, activation='relu'),
    Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

history = model.fit(
    X_train, y_train,
    epochs=10,
    batch_size=64,
    validation_split=0.1
)


Epoch 1/10


/usr/local/lib/python3.11/dist-packages/keras/src/layers/rnn/bidirectional.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


601/601 ━━━━━━━━━━━━━━━━━━━━ 11s 11ms/step - accuracy: 0.8384 - loss: 0.3654 - val_accuracy: 0.9380 - val_loss: 0.1719
Epoch 2/10
601/601 ━━━━━━━━━━━━━━━━━━━━ 5s 9ms/step - accuracy: 0.9434 - loss: 0.1582 - val_accuracy: 0.9461 - val_loss: 0.1374
Epoch 3/10
601/601 ━━━━━━━━━━━━━━━━━━━━ 11s 10ms/step - accuracy: 0.9521 - loss: 0.1313 - val_accuracy: 0.9579 - val_loss: 0.1184
Epoch 4/10
601/601 ━━━━━━━━━━━━━━━━━━━━ 5s 8ms/step - accuracy: 0.9595 - loss: 0.1129 - val_accuracy: 0.9679 - val_loss: 0.1027
Epoch 5/10
601/601 ━━━━━━━━━━━━━━━━━━━━ 7s 12ms/step - accuracy: 0.9633 - loss: 0.1060 - val_accuracy: 0.9607 - val_loss: 0.1077
Epoch 6/10
601/601 ━━━━━━━━━━━━━━━━━━━━ 7s 11ms/step - accuracy: 0.9641 - loss: 0.1007 - val_accuracy: 0.9628 - val_loss: 0.1080
Epoch 7/10
601/601 ━━━━━━━━━━━━━━━━━━━━ 10s 10ms/step - accuracy: 0.9679 - loss: 0.0920 - val_accuracy: 0.9623 - val_loss: 0.1046
Epoch 8/10
601/601 ━━━━━━━━━━━━━━━━━━━━ 8s 13ms/step - accuracy: 0.9690 - loss: 0.0903 - val_accuracy: 0.95

In [41]:
y_pred_probs = model.predict(X_test).flatten()
y_pred = (y_pred_probs > 0.5).astype(int)


print("Classification Report:\n", classification_report(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("AUC Score:", roc_auc_score(y_test, y_pred_probs))


334/334 ━━━━━━━━━━━━━━━━━━━━ 5s 11ms/step
Classification Report:
               precision    recall  f1-score   support

           0       0.98      0.97      0.97      5277
           1       0.97      0.98      0.97      5401

    accuracy                           0.97     10678
   macro avg       0.97      0.97      0.97     10678
weighted avg       0.97      0.97      0.97     10678

Confusion Matrix:
 [[5109  168]
 [ 113 5288]]
AUC Score: 0.9951877257129617


#Observation:
* We use oversampling but the problem is the time series is broken. Although it gives great result but it can not predict the trend. so we can not go forward with this.
* i see the 0 --> 33446 nos. and 1--> 442 nos. and the ratio is approx 75 : 1
this output is giving every 15 minute. so in one day 24* 4 = 96 data point.
so if we consider for 1 days and if it is found out at least single "1" in the whole day, the output will be 1 for that day. such we can balance the dataset. lets see what happen.




In [42]:
# Ensuring Timestamp is in datetime
df['Timestamp'] = pd.to_datetime(df['Timestamp'])

# Creating a 'date' column
df['date'] = df['Timestamp'].dt.date

# Daily aggregation
daily_df = df.groupby('date').agg({
    'T_Supply': 'mean',
    'T_Return': 'mean',
    'SP_Return': 'mean',
    'T_Saturation': 'mean',
    'T_Outdoor': 'mean',
    'RH_Supply': 'mean',
    'RH_Return': 'mean',
    'RH_Outdoor': 'mean',
    'Energy': 'sum',
    'Power': 'sum',
    'overheat_flag': 'max',
})

# Rename target
daily_df['target'] = daily_df['overheat_flag']
daily_df.drop(columns='overheat_flag', inplace=True)

# Now checking balance
print(daily_df['target'].value_counts())

target
0    216
1    145
Name: count, dtype: int64


In [43]:
X = daily_df.drop(columns=['target'])
y = daily_df['target']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, shuffle=False
)

model = RandomForestClassifier(random_state=42)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
y_proba = model.predict_proba(X_test)[:, 1]

print("Classification Report:\n", classification_report(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("AUC Score:", roc_auc_score(y_test, y_proba))

Classification Report:
               precision    recall  f1-score   support

           0       0.96      0.77      0.85        64
           1       0.32      0.78      0.45         9

    accuracy                           0.77        73
   macro avg       0.64      0.77      0.65        73
weighted avg       0.88      0.77      0.80        73

Confusion Matrix:
 [[49 15]
 [ 2  7]]
AUC Score: 0.875


# Observation:
* Here we got good result, but see one thing.
0 -->   216,
1 -->   145

so approx every 2 days there is a day with over heating. which is not adorable. now let do oversampling using `T-Smote`

In [44]:


#  TSMOTE function
def advanced_tsmote(X, y, minority_class=1, k=5, n_samples=None, random_state=42):
    np.random.seed(random_state)
    X = np.array(X)
    y = np.array(y)

    X_min = X[y == minority_class]
    y_min = y[y == minority_class]
    n_min = len(X_min)

    if n_samples is None:
        n_maj = len(y[y != minority_class])
        n_samples = n_maj - n_min

    X_min_flat = X_min.reshape(n_min, -1)

    from sklearn.neighbors import NearestNeighbors
    knn = NearestNeighbors(n_neighbors=k+1)
    knn.fit(X_min_flat)

    synthetic_samples = []

    for _ in range(n_samples):
        i = np.random.randint(0, n_min)
        xi = X_min[i]
        xi_flat = X_min_flat[i].reshape(1, -1)

        neighbors_idx = knn.kneighbors(xi_flat, return_distance=False)[0][1:]
        j = np.random.choice(neighbors_idx)
        xj = X_min[j]

        lambdas = np.random.uniform(0, 1, size=(xi.shape[0], 1))
        synthetic = xi + lambdas * (xj - xi)
        synthetic_samples.append(synthetic)

    X_syn = np.array(synthetic_samples)
    y_syn = np.full(X_syn.shape[0], minority_class)

    X_combined = np.concatenate([X, X_syn], axis=0)
    y_combined = np.concatenate([y, y_syn], axis=0)

    return X_combined, y_combined




In [45]:
df.head()

Timestamp   T_Supply   T_Return  SP_Return  T_Saturation  \
0 2019-10-14 22:00:00+00:00  19.859999  20.469999       18.5         19.02   
1 2019-10-14 22:15:00+00:00  19.855000  20.430000       18.5         19.02   
2 2019-10-14 22:30:00+00:00  19.850000  20.410000       18.5         19.02   
3 2019-10-14 22:45:00+00:00  19.840000  20.379999       18.5         19.08   
4 2019-10-14 23:00:00+00:00  19.830000  20.350000       18.5         19.08   

   T_Outdoor  RH_Supply  RH_Return  RH_Outdoor  Energy  Power  mode  \
0  20.299999  71.110001  58.919998        79.5     0.0    0.0  idle   
1  20.299999  71.320000  59.000000        82.0     0.0    0.0  idle   
2  20.299999  71.470001  59.109997        79.5     0.0    0.0  idle   
3  20.299999  71.439995  59.309998        77.0     0.0    0.0  idle   
4  20.299999  71.580002  59.559998        79.5     0.0    0.0  idle   

   T_Supply_roll3  overheat_flag  T_Return_lag1  Power_lag1  \
0             NaN              0            NaN         NaN   
1             NaN              0      20.469999         0.0   
2       19.855000              0      20.430000         0.0   
3       19.848333              0      20.410000         0.0   
4       19.840000              0      20.379999         0.0   

   T_Saturation_lag1  delta_T_Return  Power_roll3  hour  dayofweek        date  
0                NaN             NaN          NaN    22          0  2019-10-14  
1              19.02       -0.039999          NaN    22          0  2019-10-14  
2              19.02       -0.020000          0.0    22          0  2019-10-14  
3              19.02       -0.030001          0.0    22          0  2019-10-14  
4              19.08       -0.029999          0.0    23          0  2019-10-14

In [46]:
df.columns


Index(['Timestamp', 'T_Supply', 'T_Return', 'SP_Return', 'T_Saturation',
       'T_Outdoor', 'RH_Supply', 'RH_Return', 'RH_Outdoor', 'Energy', 'Power',
       'mode', 'T_Supply_roll3', 'overheat_flag', 'T_Return_lag1',
       'Power_lag1', 'T_Saturation_lag1', 'delta_T_Return', 'Power_roll3',
       'hour', 'dayofweek', 'date'],
      dtype='object')

In [47]:
dataset = df[ [
     'T_Return', 'SP_Return', 'T_Saturation',
       'T_Outdoor', 'RH_Supply', 'RH_Return', 'RH_Outdoor', 'Energy', 'Power','overheat_flag'
]]

In [48]:
dataset

T_Return  SP_Return  T_Saturation  T_Outdoor  RH_Supply  RH_Return  \
0      20.469999       18.5     19.020000  20.299999  71.110001  58.919998   
1      20.430000       18.5     19.020000  20.299999  71.320000  59.000000   
2      20.410000       18.5     19.020000  20.299999  71.470001  59.109997   
3      20.379999       18.5     19.080000  20.299999  71.439995  59.309998   
4      20.350000       18.5     19.080000  20.299999  71.580002  59.559998   
...          ...        ...           ...        ...        ...        ...   
33883  20.004999       20.5     19.619999  14.700000  39.020000  27.930000   
33884  19.949999       20.5     19.539999  13.700000  39.020000  28.090000   
33885  19.955000       20.5     19.420000  13.700000  39.399998  27.930000   
33886  19.920000       20.5     19.400000  13.700000  39.599998  28.039999   
33887  19.900000       20.5     19.400000  13.700000  39.599998  28.150000   

       RH_Outdoor  Energy  Power  overheat_flag  
0            79.5     0.0    0.0              0  
1            82.0     0.0    0.0              0  
2            79.5     0.0    0.0              0  
3            77.0     0.0    0.0              0  
4            79.5     0.0    0.0              0  
...           ...     ...    ...            ...  
33883        57.0     0.0    0.0              0  
33884        57.0     0.0    0.0              0  
33885        57.0     0.0    0.0              0  
33886        57.0     0.0    0.0              0  
33887        57.0     0.0    0.0              0  

[33888 rows x 10 columns]

In [49]:
# Before applying TSMOTE making the data set for trainging purpose and prediction dataset for future
train_dataset = dataset.iloc[:-1000]
test_dataset = dataset.iloc[-1000:]
train_dataset.to_csv("final_train_dataset.csv", index=False)
test_dataset.to_csv("final_test_dataset.csv", index=False)


In [50]:
train_dataset.head()

T_Return  SP_Return  T_Saturation  T_Outdoor  RH_Supply  RH_Return  \
0  20.469999       18.5         19.02  20.299999  71.110001  58.919998   
1  20.430000       18.5         19.02  20.299999  71.320000  59.000000   
2  20.410000       18.5         19.02  20.299999  71.470001  59.109997   
3  20.379999       18.5         19.08  20.299999  71.439995  59.309998   
4  20.350000       18.5         19.08  20.299999  71.580002  59.559998   

   RH_Outdoor  Energy  Power  overheat_flag  
0        79.5     0.0    0.0              0  
1        82.0     0.0    0.0              0  
2        79.5     0.0    0.0              0  
3        77.0     0.0    0.0              0  
4        79.5     0.0    0.0              0

In [51]:


features = [
     'T_Return', 'SP_Return', 'T_Saturation',
       'T_Outdoor', 'RH_Supply', 'RH_Return', 'RH_Outdoor', 'Energy', 'Power',
]

target = 'overheat_flag'

df = train_dataset

X = df[features].values
y = df[target].values

# Train-test split (no shuffle to respect time order)
X_train, X_test, y_train, y_test = train_test_split(X, y, shuffle=False, test_size=0.2)

# Applying TSmote
X_train_reshaped = X_train.reshape((X_train.shape[0], X_train.shape[1], 1))  # shape: (samples, timesteps, features=1)
X_resampled, y_resampled = advanced_tsmote(X_train_reshaped, y_train, k=5)
X_resampled = X_resampled.reshape(X_resampled.shape[0], -1)  # back to 2D

print("Before Oversampling:", Counter(y_train))
print("After  Oversampling:", Counter(y_resampled))



Before Oversampling: Counter({np.int64(0): 25948, np.int64(1): 362})
After  Oversampling: Counter({np.int64(0): 25948, np.int64(1): 25948})


In [52]:
X_train = X_resampled
y_train = y_resampled


In [53]:


# Model training
model = RandomForestClassifier(n_estimators=100, random_state=42, class_weight='balanced')
model.fit(X_resampled, y_resampled)

# Evaluation on untouched test set
y_pred = model.predict(X_test)
y_probs = model.predict_proba(X_test)[:, 1]

print("\nClassification Report Random Forest:\n", classification_report(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("AUC Score:", roc_auc_score(y_test, y_probs))



Classification Report Random Forest:
               precision    recall  f1-score   support

           0       0.99      0.99      0.99      6498
           1       0.33      0.42      0.37        80

    accuracy                           0.98      6578
   macro avg       0.66      0.71      0.68      6578
weighted avg       0.98      0.98      0.98      6578

Confusion Matrix:
 [[6429   69]
 [  46   34]]
AUC Score: 0.9638552246845182


In [54]:

# Applying XGBoost
model = XGBClassifier(n_estimators=100, random_state=42, use_label_encoder=False, eval_metric='logloss')
model.fit(X_resampled, y_resampled)

# Evaluation on untouched test set
y_pred = model.predict(X_test)
y_probs = model.predict_proba(X_test)[:, 1]

print("\nClassification Report XG Boost:\n", classification_report(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("AUC Score:", roc_auc_score(y_test, y_probs))


/usr/local/lib/python3.11/dist-packages/xgboost/core.py:158: UserWarning: [03:45:19] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)



Classification Report XG Boost:
               precision    recall  f1-score   support

           0       0.99      0.98      0.99      6498
           1       0.29      0.53      0.38        80

    accuracy                           0.98      6578
   macro avg       0.64      0.75      0.68      6578
weighted avg       0.99      0.98      0.98      6578

Confusion Matrix:
 [[6397  101]
 [  38   42]]
AUC Score: 0.9561095721760543


In [55]:
!pip install optuna xgboost


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 395.9/395.9 kB 13.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 247.0/247.0 kB 18.7 MB/s eta 0:00:00


In [56]:
print(np.isnan(X).sum(), np.isinf(X).sum())
print(np.isnan(y).sum(), np.isinf(y).sum())


0 0
0 0


In [57]:
# Use resampled data
X = np.array(X_resampled)
y = np.array(y_resampled)

# Handling NaN and Inf values
print("Before cleanup:", np.isnan(X).sum(), np.isinf(X).sum())

# Drop rows with NaNs in X
mask = ~np.isnan(X).any(axis=1)
X = X[mask]
y = y[mask]

print("After cleanup:", np.isnan(X).sum(), np.isinf(X).sum())

# Scaling features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Reshape for LSTM: (samples, timesteps, features)
X_lstm = X_scaled.reshape((X_scaled.shape[0], X_scaled.shape[1], 1))

# Encode target if multi-class
num_classes = len(np.unique(y))
if num_classes > 2:
    y = to_categorical(y)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X_lstm, y, test_size=0.2, random_state=42, stratify=y if num_classes > 2 else None
)

# Build LSTM model
model = Sequential()
model.add(LSTM(64, input_shape=(X_train.shape[1], 1), return_sequences=False))
model.add(Dropout(0.2))

if num_classes > 2:
    model.add(Dense(num_classes, activation='softmax'))
    loss_fn = 'categorical_crossentropy'
else:
    model.add(Dense(1, activation='sigmoid'))
    loss_fn = 'binary_crossentropy'

model.compile(optimizer='adam', loss=loss_fn, metrics=['accuracy'])

# Train the model
history = model.fit(
    X_train, y_train,
    epochs=20,
    batch_size=32,
    validation_data=(X_test, y_test),
    verbose=1
)

# Evaluate
y_pred = model.predict(X_test)

if num_classes > 2:
    y_pred_classes = np.argmax(y_pred, axis=1)
    y_true_classes = np.argmax(y_test, axis=1)
else:
    y_pred_classes = (y_pred > 0.5).astype("int32").flatten()
    y_true_classes = y_test

print("Classification Report:\n", classification_report(y_true_classes, y_pred_classes))
print("Confusion Matrix:\n", confusion_matrix(y_true_classes, y_pred_classes))


Before cleanup: 0 0
After cleanup: 0 0
Epoch 1/20


/usr/local/lib/python3.11/dist-packages/keras/src/layers/rnn/rnn.py:200: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


1298/1298 ━━━━━━━━━━━━━━━━━━━━ 24s 16ms/step - accuracy: 0.8708 - loss: 0.3374 - val_accuracy: 0.9397 - val_loss: 0.1737
Epoch 2/20
1298/1298 ━━━━━━━━━━━━━━━━━━━━ 33s 10ms/step - accuracy: 0.9402 - loss: 0.1731 - val_accuracy: 0.9482 - val_loss: 0.1482
Epoch 3/20
1298/1298 ━━━━━━━━━━━━━━━━━━━━ 10s 8ms/step - accuracy: 0.9450 - loss: 0.1557 - val_accuracy: 0.9496 - val_loss: 0.1361
Epoch 4/20
1298/1298 ━━━━━━━━━━━━━━━━━━━━ 25s 12ms/step - accuracy: 0.9461 - loss: 0.1434 - val_accuracy: 0.9557 - val_loss: 0.1233
Epoch 5/20
1298/1298 ━━━━━━━━━━━━━━━━━━━━ 17s 9ms/step - accuracy: 0.9533 - loss: 0.1268 - val_accuracy: 0.9576 - val_loss: 0.1153
Epoch 6/20
1298/1298 ━━━━━━━━━━━━━━━━━━━━ 11s 9ms/step - accuracy: 0.9580 - loss: 0.1170 - val_accuracy: 0.9664 - val_loss: 0.0989
Epoch 7/20
1298/1298 ━━━━━━━━━━━━━━━━━━━━ 20s 8ms/step - accuracy: 0.9631 - loss: 0.1047 - val_accuracy: 0.9647 - val_loss: 0.0968
Epoch 8/20
1298/1298 ━━━━━━━━━━━━━━━━━━━━ 22s 9ms/step - accuracy: 0.9653 - loss: 0.0980 - 

### It is seem that LSTM does not perform well. XGBoost perform better at the moment. So lets go forward with XGBoost

In [58]:
X_resampled

array([[20.469999  , 18.5       , 19.02      , ..., 79.5       ,
         0.        ,  0.        ],
       [20.43      , 18.5       , 19.02      , ..., 82.        ,
         0.        ,  0.        ],
       [20.41      , 18.5       , 19.02      , ..., 79.5       ,
         0.        ,  0.        ],
       ...,
       [19.72356923, 22.5       , 19.199999  , ..., 94.        ,
         6.        ,  2.51292502],
       [18.58954738, 22.5       , 17.24695731, ..., 87.        ,
        12.        ,  4.91120026],
       [18.10380572, 22.5       , 17.79471073, ..., 86.        ,
        13.        ,  5.004     ]])

In [59]:
y_resampled

array([0, 0, 0, ..., 1, 1, 1])

In [61]:
import optuna

X = X_resampled
y = y_resampled

# Apply MinMaxScaler BEFORE train-test split
scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X)

# Split the data
X_train, X_val, y_train, y_val = train_test_split(
    X_scaled, y, test_size=0.2, random_state=35, stratify=y
)

# Optuna objective function
def objective(trial):
    param = {
        'n_estimators': trial.suggest_int('n_estimators', 100, 500),
        'learning_rate': trial.suggest_float('learning_rate', 0.01, 0.3),
        'max_depth': trial.suggest_int('max_depth', 3, 10),
        'subsample': trial.suggest_float('subsample', 0.5, 1.0),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.5, 1.0),
        'gamma': trial.suggest_float('gamma', 0, 0.5),
        'scale_pos_weight': trial.suggest_float('scale_pos_weight', 1, 20),
        'use_label_encoder': False,
        'eval_metric': 'logloss'
    }

    model = XGBClassifier(**param)
    model.fit(X_train, y_train)

    y_pred = model.predict(X_val)
    return f1_score(y_val, y_pred)

# Run Optuna study
study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=200)

# Best parameters
best_params = study.best_params
print("Best Trial Parameters:", best_params)

# Final model with best parameters
final_model = XGBClassifier(
    **best_params,
    use_label_encoder=False,
    eval_metric='logloss'
)
final_model.fit(X_train, y_train)

# Evaluation
y_pred = final_model.predict(X_val)
print("Classification Report:\n", classification_report(y_val, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_val, y_pred))
print("AUC Score:", roc_auc_score(y_val, final_model.predict_proba(X_val)[:, 1]))


[I 2025-07-21 03:53:45,136] A new study created in memory with name: no-name-aac35d3d-3686-4a75-ad29-a69a28a9d324
/usr/local/lib/python3.11/dist-packages/xgboost/core.py:158: UserWarning: [03:53:45] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)
[I 2025-07-21 03:53:48,814] Trial 0 finished with value: 0.99149058227364 and parameters: {'n_estimators': 252, 'learning_rate': 0.06971699587680788, 'max_depth': 8, 'subsample': 0.7675035369112952, 'colsample_bytree': 0.7981105333562567, 'gamma': 0.2508897152351946, 'scale_pos_weight': 7.888479426531065}. Best is trial 0 with value: 0.99149058227364.
/usr/local/lib/python3.11/dist-packages/xgboost/core.py:158: UserWarning: [03:53:48] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)
[I 2025-07-21 03:53:49,678] Trial 1 finished with value: 0.9900706511361467 and parameters: {'n_estimators':

Best Trial Parameters: {'n_estimators': 408, 'learning_rate': 0.27359082963308273, 'max_depth': 10, 'subsample': 0.7759961590167008, 'colsample_bytree': 0.5119088159506807, 'gamma': 0.028481668629778693, 'scale_pos_weight': 1.1332352118338844}
Classification Report:
               precision    recall  f1-score   support

           0       1.00      0.99      1.00      5190
           1       0.99      1.00      1.00      5190

    accuracy                           1.00     10380
   macro avg       1.00      1.00      1.00     10380
weighted avg       1.00      1.00      1.00     10380

Confusion Matrix:
 [[5156   34]
 [   7 5183]]
AUC Score: 0.9998639001191709


In [62]:
best_params = study.best_params
best_params

{'n_estimators': 408,
 'learning_rate': 0.27359082963308273,
 'max_depth': 10,
 'subsample': 0.7759961590167008,
 'colsample_bytree': 0.5119088159506807,
 'gamma': 0.028481668629778693,
 'scale_pos_weight': 1.1332352118338844}

In [63]:

best_params = study.best_params

# Update best_params for binary classification
best_params.update({
    'use_label_encoder': False,
    'eval_metric': 'logloss',
    'objective': 'binary:logistic',
    'random_state': 35
})

# Scale  using MinMaxScaler
scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X)

# Save the scaler
joblib.dump(scaler, "scaler.pkl")

# Train-test split
X_train, X_val, y_train, y_val = train_test_split(
    X_scaled, y, test_size=0.2, random_state=35, stratify=y
)

# Train the model
final_model = XGBClassifier(**best_params)
final_model.fit(X_train, y_train)

# Save the trained model
joblib.dump(final_model, "xgb_final_model.pkl")

# Make predictions
y_pred = final_model.predict(X_val)
y_proba = final_model.predict_proba(X_val)[:, 1]

# Evaluate
print("Classification Report:\n", classification_report(y_val, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_val, y_pred))
print("F1 Score:", f1_score(y_val, y_pred))
print("AUC Score:", roc_auc_score(y_val, y_proba))


/usr/local/lib/python3.11/dist-packages/xgboost/core.py:158: UserWarning: [04:00:28] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)


Classification Report:
               precision    recall  f1-score   support

           0       1.00      0.99      1.00      5190
           1       0.99      1.00      1.00      5190

    accuracy                           1.00     10380
   macro avg       1.00      1.00      1.00     10380
weighted avg       1.00      1.00      1.00     10380

Confusion Matrix:
 [[5149   41]
 [   6 5184]]
F1 Score: 0.9954872779644743
AUC Score: 0.9998362977565424


In [64]:
import joblib
joblib.dump(final_model, 'xgb_final_model.pkl')


['xgb_final_model.pkl']

In [65]:

from datetime import timedelta

def predict_overheat(model, recent_data, timestep, timestep_unit):
    """
    Predict if overheating will occur in the future time window.

    Args:
    - model: trained XGBoost model
    - recent_data: DataFrame containing most recent data rows (15-min interval)
    - timestep: integer (how many units forward to predict)
    - timestep_unit: "minute", "hour", "day"

    Returns:
    - prediction: 0 or 1 (1 means overheat will occur)
    - expected_overheat_time: estimated datetime of overheating (if any)
    """

    interval_minutes = 15  # dataset frequency
    future_minutes = timestep

    if timestep_unit == "hour":
        future_minutes = timestep * 60
    elif timestep_unit == "day":
        future_minutes = timestep * 24 * 60

    n_steps = future_minutes // interval_minutes

    # Ensure enough data
    if len(recent_data) < n_steps:
        return "Not enough data", None

    X_input = recent_data[-n_steps:]  # most recent 'n_steps' rows
    preds = model.predict(X_input)

    if 1 in preds:
        overheat_index = np.where(preds == 1)[0][0]
        overheat_time = X_input.index[overheat_index]
        return 1, overheat_time
    else:
        return 0, None


In [67]:
import pandas as pd
import numpy as np
import joblib
from sklearn.utils import shuffle

# Load your model and MinMax scaler
model = joblib.load("xgb_final_model.pkl")
scaler = joblib.load("scaler.pkl")

# Load dataset
data = pd.read_csv("final_test_dataset.csv")

# Add new rows (same number of features as input, without output column)
new_rows = [
    [22.059999, 19.5, 20.219999, 14.3, 36.160000, 26.029999, 58.0, 14.0, 5.064],
    [20.754999, 19.5, 18.920000, 9.1, 49.039997, 39.160000, 93.0, 13.0, 5.136],
    [19.365000, 20.5, 18.660000, 7.1, 39.270000, 34.309998, 87.0, 13.0, 4.956],
    [20.660000, 20.5, 17.520000, 12.3, 41.299999, 34.169998, 81.0, 9.0, 4.968],
    [20.330000, 20.5, 17.139999, 8.1, 39.040001, 33.130001, 80.0, 10.0, 4.968],
]

# Drop output column
recent_data = data.iloc[:, :-1].copy()

# Append new rows
new_df = pd.DataFrame(new_rows, columns=recent_data.columns)
recent_data = pd.concat([recent_data, new_df], ignore_index=True)

# Shuffle the dataset
recent_data = shuffle(recent_data, random_state=42).reset_index(drop=True)

# Start infinite loop
while True:
    print("\n🔁 Predict Overheat - Type 'exit' to quit")

    time_unit = input("Enter time unit (minute/hour/day): ").strip().lower()
    if time_unit == "exit":
        print("👋 Exiting system.")
        break

    try:
        time_value = int(input(f"Enter number of {time_unit}s: ").strip())
    except ValueError:
        print("❌ Invalid number. Try again.")
        continue

    # Convert time to number of 15-minute steps
    if time_unit == "minute":
        steps = time_value // 15
    elif time_unit == "hour":
        steps = (time_value * 60) // 15
    elif time_unit == "day":
        steps = (time_value * 24 * 60) // 15
    else:
        print("❌ Invalid time unit! Use minute/hour/day.")
        continue

    if steps <= 0:
        print("❌ Time value too low to form steps.")
        continue
    elif steps > len(recent_data):
        print(f"⚠️ Not enough data for {steps} steps. Using last {len(recent_data)} steps instead.")
        steps = len(recent_data)

    # Select last `steps` rows
    recent_steps = recent_data.tail(steps)

    # Drop unwanted columns (like "index" if it exists)
    X = recent_steps.drop(columns=["index"], errors="ignore")

    # Scale features
    X_scaled = scaler.transform(X)

    # Predict
    y_pred = model.predict(X_scaled)

    if np.any(y_pred == 1):
        print("⚠️ System will OVERHEAT in this interval!")
    else:
        print("✅ System will NOT overheat in this interval.")



🔁 Predict Overheat - Type 'exit' to quit
Enter time unit (minute/hour/day): minute
Enter number of minutes: 30


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but MinMaxScaler was fitted without feature names
  warnings.warn(


✅ System will NOT overheat in this interval.

🔁 Predict Overheat - Type 'exit' to quit
Enter time unit (minute/hour/day): hour
Enter number of hours: 6


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but MinMaxScaler was fitted without feature names
  warnings.warn(


✅ System will NOT overheat in this interval.

🔁 Predict Overheat - Type 'exit' to quit
Enter time unit (minute/hour/day): day
Enter number of days: 1


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but MinMaxScaler was fitted without feature names
  warnings.warn(


✅ System will NOT overheat in this interval.

🔁 Predict Overheat - Type 'exit' to quit
Enter time unit (minute/hour/day): day
Enter number of days: 3


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but MinMaxScaler was fitted without feature names
  warnings.warn(


✅ System will NOT overheat in this interval.

🔁 Predict Overheat - Type 'exit' to quit
Enter time unit (minute/hour/day): day
Enter number of days: 4


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but MinMaxScaler was fitted without feature names
  warnings.warn(


✅ System will NOT overheat in this interval.

🔁 Predict Overheat - Type 'exit' to quit
Enter time unit (minute/hour/day): day
Enter number of days: 5


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but MinMaxScaler was fitted without feature names
  warnings.warn(


⚠️ System will OVERHEAT in this interval!

🔁 Predict Overheat - Type 'exit' to quit
Enter time unit (minute/hour/day): exit
👋 Exiting system.


# Future work:
* Buildng maore model with various Algorith like Random Forest, LSTM, Transformer and so on. Everyday the all models will run and which will give best result, will be picked for that day.
* Making more advanced using GenAI, Agent AI.
* Building a pure software.